### PR TITLE
Implemented JSONRPC 2.0 batch calls as per spec

### PR DESCRIPTION
**Pull request by JustusW**
Implemented via Pyramid Tweens as suggested in TODO. All tests in test_jsonrpc.TestJSONRPCIntegration have been updated to work both batched and unbatched.

Testprotocol:
(acpdev)juw@juw:~/Projects/pyramid_rpc/pyramid_rpc/tests$ python -m unittest test_jsonrpc
...No handlers could be found for logger "pyramid_rpc.jsonrpc"
.....................................
----------------------------------------------------------------------
Ran 40 tests in 0.218s

OK

### CODE REVIEW

File: pyramid_rpc/jsonrpc.py [hunk 1]
```diff
@@ -160,6 +160,44 @@ def __call__(self, context, request):
...
+            # Keep the last response object and extract the json
+            # from the body.
+            response = self.handler(request)
+            response_body.append(json.loads(response.body))
+
+        response_body = [result for result in response_body if result != '']
+        if len(response_body) == 0:
+            # If no response objects are available the JSONRPC Specification forbids
+            # sending an empty array. An empty string is sent instead.
+            response.body = '""'
```
Comment (rsommer) [inline pyramid_rpc/jsonrpc.py]: The current RFC https://tools.ietf.org/html/rfc7159 does no longer force JSON documents to start with either an array or an object. A string, a number or a boolean value are also valid (see "Examples" section in RFC)

**Comment by JustusW** [comment]
Hello again, I've just added the over=EXCVIEW setting you were talking about. Sadly the 2.0 Specs are not totally explicit on this topic, but I believe it would be correct to return a separate response object even if a sub request fails.
Is it possible to merge this sometime soon?

**Comment by inklesspen** [comment]
We really want this functionality at Axial; anything I can do to help get this in?

File: pyramid_rpc/jsonrpc.py [hunk 2]
```diff
@@ -160,6 +160,44 @@ def __call__(self, context, request):
...
+    def __init__(self, handler, registry):
+        self.handler = handler
+        self.registry = registry
+
+    def __call__(self, request):
+        # Check if it is a JSON request and a list.
+        try:
+            json_body = request.json_body[:]
+        except (ValueError, TypeError):
+            return self.handler(request)
```
Comment (inklesspen) [inline pyramid_rpc/jsonrpc.py]: Frankly I would recommend splitting up this block; there's two questions at hand, is this body valid JSON and is this body an iterable. the code will be more readable with two try/except blocks.

File: pyramid_rpc/jsonrpc.py (hunk 1 above)
Comment (inklesspen) [inline pyramid_rpc/jsonrpc.py]: [If there are no Response objects contained within the Response array as it is to be sent to the client, the server MUST NOT return an empty Array and should return nothing at all.](http://www.jsonrpc.org/specification#batch)

Comment (inklesspen) [inline pyramid_rpc/jsonrpc.py]: (Personally I read that as a non-json body containing no bytes, not a JSON body consisting of `""`.)

**Comment by mmerickel** [comment]
So I just spent a while bringing pyramid-rpc back into the modern world, fixing the test runner and docs stuff. Now I've finally got to dealing with this PR. I have several concerns.

1) Tests fail on py3. I've fixed most of these.
2) The tween affects every request coming into the pyramid app. This tween needs to find a way to affect only requests where the endpoint matches one of the registered endpoints to avoid running the json deserializer on every request. If this isn't clear we can always go into more detail. (`registry.rpc_endpoints`).
3) I'm fine with the empty string as a response probably but I'd like someone to double check with the spec and verify this behavior versus an empty array. I'd lean toward `null` instead of an empty array personally.

I pushed a new branch jsonrpc-batch-support that new PRs can be submitted against if anyone can help with these issues.

**Pull request closed without merging**